File: 560-SubarraySumEqualsK/560-SubarraySumEqualsK.py
# Last updated: 8/19/2025, 12:44:49 AM
class Solution(object):
    def subarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        waysToReach = defaultdict(int)
        waysToReach[0] = 1
        running = 0
        ans = 0

        for num in nums:
            running += num
            ans += waysToReach[running - k]
            waysToReach[running] += 1
        
        return ans

File: 560-SubarraySumEqualsK/560-SubarraySumEqualsK.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        HashMap<Integer, Integer> prefixSum = new HashMap<>();
        prefixSum.put(0, 1);
        int count = 0;
        int curSum = 0;

        for (int num: nums) {
            curSum += num;

            count += prefixSum.getOrDefault(curSum - k, 0);

            prefixSum.put(curSum, prefixSum.getOrDefault(curSum, 0) + 1);
        }

        return count;
    }
}